src/init.cpp

```diff
@@ -1686,27 +1686,19 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
                 break;
             }
 
-            bool failed_rewind{false};
-            // Can't hold cs_main while calling RewindBlockIndex, so retrieve the relevant
-            // chainstates beforehand.
-            for (CChainState* chainstate : WITH_LOCK(::cs_main, return chainman.GetAll())) {
-                if (!fReset) {
-                    // Note that RewindBlockIndex MUST run even if we're about to -reindex-chainstate.
-                    // It both disconnects blocks based on the chainstate, and drops block data in
-                    // BlockIndex() based on lack of available witness data.
-                    uiInterface.InitMessage(_("Rewinding blocks...").translated);
-                    if (!chainstate->RewindBlockIndex(chainparams)) {
-                        strLoadError = _(
-                            "Unable to rewind the database to a pre-fork state. "
-                            "You will need to redownload the blockchain");
-                        failed_rewind = true;
-                        break; // out of the per-chainstate loop
-                    }
-                }
-            }
 
-            if (failed_rewind) {
-                break; // out of the chainstate activation do-while
+            if (!fReset) {
+                uiInterface.InitMessage(_("Checking if redownload is needed...").translated);
+                LOCK(cs_main);
+                auto chainstates{chainman.GetAll()};
+                if (std::any_of(chainstates.begin(), chainstates.end(),
+                                [&chainparams](CChainState* cs) EXCLUSIVE_LOCKS_REQUIRED(cs_main) { return cs->NeedsRedownload(chainparams); })) {
+                    auto chainstate_dir = GetDataDir() / CHAINSTATE_DIRNAME;
+                    strLoadError = strprintf(_(
+                        "Segwit blocks are insufficiently validated. "
+                        "Please delete blocks dir(%s) and chainstate dir(%s) and restart."), GetBlocksDir(), chainstate_dir);
```

Comment:
A `-reindex` should do the same with less effort?

Reply:
When a pre-segwit node upgrades to segwit, it has no way to sufficiently validate to the "new" consensus rules from the blocks on disk since witness data was never relayed to it. Help for `-reindex` says "Rebuild chain state and block index from the blk*.dat files on disk". Looking at the code also does not suggest to me that `-reindex` would cause the node to re-download.

Am I missing something?

Reply:
Ah, I think this is true. `-reindex` will cause the node to try to rebuild the chain state from the block files, which will work only as far as the segwit activation height if the blocks after that are serialized without witness. The node will then try to sync by downloading all blocks from peers as normal. The test passes with the following change:

```diff
diff --git a/test/functional/p2p_segwit.py b/test/functional/p2p_segwit.py
index 219c15f673..76d8967a41 100755
--- a/test/functional/p2p_segwit.py
+++ b/test/functional/p2p_segwit.py
@@ -1973,15 +1973,8 @@ class SegWitTest(BitcoinTestFramework):
         with self.nodes[2].assert_debug_log(expected_msgs=["Segwit blocks are insufficiently validated. Please delete blocks dir and chain state dir and restart."], timeout=10):
             self.nodes[2].start(["-segwitheight={}".format(SEGWIT_HEIGHT)])
 
-        # As directed, the user deletes the blocks and chainstate directories and restarts the node
-        datadir = get_datadir_path(self.options.tmpdir, 2)
-        blocks_dir = os.path.join(datadir, "regtest", "blocks")
-        chainstate_dir = os.path.join(datadir, "regtest", "chainstate")
-        shutil.rmtree(blocks_dir)
-        shutil.rmtree(chainstate_dir)
-
-        self.start_node(2, extra_args=["-segwitheight={}".format(SEGWIT_HEIGHT)])
-        assert_equal(self.nodes[2].getblockcount(), 0)
+        self.start_node(2, extra_args=["-reindex=1", "-segwitheight={}".format(SEGWIT_HEIGHT)])
+        assert_equal(self.nodes[2].getblockcount(), SEGWIT_HEIGHT - 1)
         self.connect_nodes(0, 2)
 
         # We reconnect more than 100 blocks, give it plenty of time
```

There are a bunch of `AcceptBlock` errors since the blocks after height SEGWIT_HEIGHT aren't serialized properly:

```
node2 2021-02-16T10:05:38.943607Z [loadblk] ERROR: AcceptBlock: bad-witness-nonce-size, ContextualCheckBlock : invalid witness reserved value size 
 node2 2021-02-16T10:05:38.943663Z [loadblk] ERROR: AcceptBlock: bad-witness-nonce-size, ContextualCheckBlock : invalid witness reserved value size 
 node2 2021-02-16T10:05:38.943712Z [loadblk] ERROR: AcceptBlock: bad-witness-nonce-size, ContextualCheckBlock : invalid witness reserved value size 
[...]
```

and when the node connects to its peers, it downloads and syncs to the tip, validating segwit rules as it goes.

One downside is that the old blocks are retained on disk and never deleted. However, that was the case before this PR as well.

Reply:
Ah, what I wasn't thinking about is that it is normal for the block _index_ to not be a 1:1 representation of the blk*.dat files. It happens all the time with reorgs, etc. So, upon `-reindex` node won't be able to validate `[SEGWIT_HEIGHT, consensusTip]` blocks, will have `tip == SEGWIT_HEIGHT - 1`. It will then proceed to sync headers and therefore re-download `[SEGWIT_HEIGHT, consensusTip]`.

Thank you for the great suggestion @MarcoFalke! Makes for much simpler code. Updated.